### PR TITLE
Feature/add reset argument

### PR DESCRIPTION
**Pull request by Joris-JW**
Closes PM-337

## Description
If the IMC's are not  closed/started correctly a problem could occur which requires a reset of the IMC's. Within the current setup the easy motion studio laptop has to be attached and used in order to do this. With this PR an argument has been added to the hardware_interface launch file to reset all the IMC's using the 2080 SDO register.

## Changes
* Added reset argument to hardware_interface launch file
* Added reset function

### CODE REVIEW

File: march_hardware/src/MarchRobot.cpp [hunk 1]
```diff
@@ -47,6 +47,16 @@ void MarchRobot::startEtherCAT()
     return;
   }
   ethercatMaster.start(this->jointList);
+
+  if (do_reset_imc)
+  {
+    ROS_INFO("Resetting all IMC");
```
Comment (Roelemans) [inline march_hardware/src/MarchRobot.cpp]:
```suggestion
    ROS_INFO("Resetting all IMCs");
```

File: march_hardware/include/march_hardware/MarchRobot.h [hunk 2]
```diff
@@ -40,7 +40,9 @@ class MarchRobot
   MarchRobot(MarchRobot&&) = delete;
   MarchRobot& operator=(MarchRobot&&) = delete;
 
-  void startEtherCAT();
+  void resetIMC();
```
Comment (Roelemans) [inline march_hardware/include/march_hardware/MarchRobot.h]:
```suggestion
  void resetIMotionCubes();
```

File: march_hardware/src/MarchRobot.cpp (hunk 1 above)
Comment (Roelemans) [reply march_hardware/src/MarchRobot.cpp]: You have this comment twice now: here and in resetIMC(). I would remove it here.

File: march_hardware/src/MarchRobot.cpp [hunk 3]
```diff
@@ -60,6 +70,15 @@ void MarchRobot::stopEtherCAT()
   ethercatMaster.stop();
 }
 
+void MarchRobot::resetIMC()
+{
+  ROS_INFO("Resetting all IMC on initialization");
```
Comment (Roelemans) [inline march_hardware/src/MarchRobot.cpp]:
```suggestion
  ROS_INFO("Resetting all IMCs");
```
Its not really relevant whether it is during initialization or not

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 4]
```diff
@@ -24,7 +24,7 @@ using joint_limits_interface::PositionJointSoftLimitsHandle;
 using joint_limits_interface::SoftJointLimits;
 using march::Joint;
 
-MarchHardwareInterface::MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot)
+MarchHardwareInterface::MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot, bool use_reset_imc)
```
Comment (Roelemans) [inline march_hardware_interface/src/march_hardware_interface.cpp]:
```suggestion
MarchHardwareInterface::MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot, bool reset_imc)
```

File: march_hardware_interface/include/march_hardware_interface/march_hardware_interface.h [hunk 5]
```diff
@@ -34,7 +34,7 @@ using RtPublisherPtr = std::unique_ptr<realtime_tools::RealtimePublisher<T>>;
 class MarchHardwareInterface : public hardware_interface::RobotHW
 {
 public:
-  explicit MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot);
+  explicit MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot, bool use_reset_imc);
```
Comment (Roelemans) [inline march_hardware_interface/include/march_hardware_interface/march_hardware_interface.h]:
```suggestion
  explicit MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot, bool reset_imc);
```
I think it is fine to just use reset_imc everywhere?

File: march_hardware_interface/src/march_hardware_interface_node.cpp [hunk 6]
```diff
@@ -21,10 +21,13 @@ int main(int argc, char** argv)
   AllowedRobot selected_robot = AllowedRobot(argv[1]);
   ROS_INFO_STREAM("Selected robot: " << selected_robot);
 
+  bool use_reset_imc = ros::param::param<bool>("~reset_imc", 1);
```
Comment (Roelemans) [inline march_hardware_interface/src/march_hardware_interface_node.cpp]:
```suggestion
  bool reset_imc = ros::param::param<bool>("~reset_imc", 1);
```

File: march_hardware_interface/src/march_hardware_interface_node.cpp [hunk 7]
```diff
@@ -21,10 +21,13 @@ int main(int argc, char** argv)
...
+  ROS_INFO_STREAM("Use the reset of the IMC: " << use_reset_imc);
+
   spinner.start();
 
   HardwareBuilder builder(selected_robot);
-  MarchHardwareInterface march(builder.createMarchRobot());
+  MarchHardwareInterface march(builder.createMarchRobot(), use_reset_imc);
```
Comment (Roelemans) [inline march_hardware_interface/src/march_hardware_interface_node.cpp]:
```suggestion
  MarchHardwareInterface march(builder.createMarchRobot(), reset_imc);
```

**Review by Roelemans — CHANGES_REQUESTED** [review]

File: march_hardware/include/march_hardware/MarchRobot.h [hunk 8]
```diff
@@ -40,7 +40,7 @@ class MarchRobot
   MarchRobot(MarchRobot&&) = delete;
   MarchRobot& operator=(MarchRobot&&) = delete;
 
-  void resetIMC();
+  void resetImotionCubes();
```
Comment (Roelemans) [inline march_hardware/include/march_hardware/MarchRobot.h]:
```suggestion
  void resetIMotionCubes();
```
iMotionCube is written with capital M everywhere else :sweat_smile: sorry

**Review by Roelemans — CHANGES_REQUESTED** [review]

**Review by Roelemans — DISMISSED** [review]
Good luck with the merge conflicts

File: march_hardware/include/march_hardware/MarchRobot.h [hunk 9]
```diff
@@ -40,7 +40,9 @@ class MarchRobot
   MarchRobot(MarchRobot&&) = delete;
   MarchRobot& operator=(MarchRobot&&) = delete;
 
-  void startEtherCAT();
+  void resetIMotionCubes();
+
+  void startEtherCAT(bool do_reset_imc);
```
Comment (Olavhaasie) [inline march_hardware/include/march_hardware/MarchRobot.h]:
```suggestion
  void startEtherCAT(bool reset_imc);
```

File: march_hardware_interface/launch/hardware.launch [hunk 10]
```diff
@@ -1,6 +1,14 @@
 <launch>
     <arg name="robot" default="march4" doc="The robot to run. Can be: march3, march4, test_joint_linear, test_joint_rotational."/>
 
+    <arg name="reset_imc" default="false" doc="Reset the IMC if this argument is set to true"/>
+
+    <arg name="control_loop_multiplier" default="1"
+         doc="Multiplier for the duration of the control loop cycle time compared
+              to the ethercat cycle time. This must be a positive integer and
+              always assures that the control loop cycle time is a multiple of
+              the ethercat cycle time."/>
```
Comment (Olavhaasie) [inline march_hardware_interface/launch/hardware.launch]: I think this is a merge artifact and should be removed

File: march_hardware_interface/launch/hardware.launch [hunk 11]
```diff
@@ -20,6 +28,10 @@
                 args="$(arg robot)"
                 output="screen"
                 required="true"
-        />
+        >
+            <param name="control_loop_multiplier" value="$(arg control_loop_multiplier)"/>
```
Comment (Olavhaasie) [inline march_hardware_interface/launch/hardware.launch]: Same here

File: march_hardware_interface/src/march_hardware_interface_node.cpp [hunk 12]
```diff
@@ -21,10 +21,13 @@ int main(int argc, char** argv)
   AllowedRobot selected_robot = AllowedRobot(argv[1]);
   ROS_INFO_STREAM("Selected robot: " << selected_robot);
 
+  bool reset_imc = ros::param::param<bool>("~reset_imc", 1);
```
Comment (Olavhaasie) [inline march_hardware_interface/src/march_hardware_interface_node.cpp]: Is this 1 the default value? Since this is a `bool` it would be better to use `true`, which is the same as 1. Actually, I think this default value should be `false`, since most of the time you don't want to reset and it agrees with the launch file.

File: march_hardware_interface/src/march_hardware_interface_node.cpp [hunk 13]
```diff
@@ -21,10 +21,13 @@ int main(int argc, char** argv)
...
+  ROS_INFO_STREAM("Use the reset of the IMC: " << reset_imc);
```
Comment (Olavhaasie) [inline march_hardware_interface/src/march_hardware_interface_node.cpp]: The resetting is also logged when it happens, so this seems redundant and only necessary for debugging.

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 14]
```diff
@@ -36,8 +36,9 @@ MarchHardwareInterface::MarchHardwareInterface(std::unique_ptr<march::MarchRobot
       this->joint_names_.push_back(urdf_joint.first);
     }
   }
-
   this->num_joints_ = this->joint_names_.size();
+
+  this->reset_imc_ = reset_imc;
```
Comment (Olavhaasie) [inline march_hardware_interface/src/march_hardware_interface.cpp]: It is better to initialize this variable in the initializer list of the constructor, so the same as how `march_robot_` and `has_power_distribution_board_` are initialized. See https://en.cppreference.com/w/cpp/language/initializer_list

**Review by Olavhaasie — CHANGES_REQUESTED** [review]

File: march_hardware_interface/src/march_hardware_interface.cpp (hunk 14 above)
Comment (Joris-JW) [reply march_hardware_interface/src/march_hardware_interface.cpp]: nice touch

File: march_hardware_interface/src/march_hardware_interface_node.cpp (hunk 12 above)
Comment (Joris-JW) [reply march_hardware_interface/src/march_hardware_interface_node.cpp]: Indeed, thought i had already changed this. Seems I was not thinking clearly during this PR. wont happen again

File: march_hardware/include/march_hardware/Joint.h [hunk 15]
```diff
@@ -40,11 +40,14 @@ class Joint
...
 
+  void resetIMotionCube();
+
   /* Delete move assignment since string cannot be move assigned */
   Joint(Joint&&) = default;
   Joint& operator=(Joint&&) = delete;
 
   void initialize(int cycle_time);
+
   void prepareActuation();
```
Comment (Roelemans) [inline march_hardware/include/march_hardware/Joint.h]: I think you should remove this newline since these two methods are related

File: march_hardware/src/Joint.cpp [hunk 16]
```diff
@@ -55,7 +55,16 @@ void Joint::prepareActuation()
...
+void Joint::resetIMotionCube()
+{
+  if (hasIMotionCube())
+  {
+    this->imc_->resetIMotionCube();
+  }
+}
+
 void Joint::actuateRad(double target_position)
+
```
Comment (Roelemans) [inline march_hardware/src/Joint.cpp]: Remove this one too

File: march_hardware_interface/launch/hardware.launch [hunk 17]
```diff
@@ -1,6 +1,9 @@
 <launch>
     <arg name="robot" default="march4" doc="The robot to run. Can be: march3, march4, test_joint_linear, test_joint_rotational."/>
 
+    <arg name="reset_imc" default="false" doc="Reset the IMC if this argument is set to true"/>
+
+
```
Comment (Roelemans) [inline march_hardware_interface/launch/hardware.launch]: Empty line at line 3 and 5 here too :innocent:  (so there is only one newline after the reset_imc arg)

File: march_hardware_interface/launch/hardware.launch [hunk 18]
```diff
@@ -20,6 +23,9 @@
                 args="$(arg robot)"
                 output="screen"
                 required="true"
-        />
+        >
+            <param name="reset_imc" value="$(arg reset_imc)"/>
+
```
Comment (Roelemans) [inline march_hardware_interface/launch/hardware.launch]: Remove this newline too :innocent:

**Review by Roelemans — CHANGES_REQUESTED** [review]
Looks good :cat:

File: march_hardware/src/IMotionCube.cpp [hunk 19]
```diff
@@ -322,6 +322,13 @@ void IMotionCube::goToOperationEnabled()
   this->goToTargetState(IMotionCubeTargetState::OPERATION_ENABLED);
 }
 
+void IMotionCube::resetIMotionCube()
+{
+  this->setControlWord(0);
+  ROS_DEBUG("Slave: %d, Try to reset IMC", this->slaveIndex);
+  sdo_bit16(slaveIndex, 0x2080, 0, 1);
```
Comment (Olavhaasie) [inline march_hardware/src/IMotionCube.cpp]: You use `this->` in the line above so you should do it here as well. (Actually we use it everywhere)
```suggestion
  sdo_bit16(this->slaveIndex, 0x2080, 0, 1);
```

File: march_hardware/include/march_hardware/IMotionCube.h [hunk 20]
```diff
@@ -64,6 +64,8 @@ class IMotionCube : public Slave
   void goToTargetState(IMotionCubeTargetState target_state);
   virtual void goToOperationEnabled();
 
+  void resetIMotionCube();
```
Comment (Olavhaasie) [inline march_hardware/include/march_hardware/IMotionCube.h]: The class is already called `IMotionCube` so it would be weird to call this method `resetIMotionCube`, since that is what it is. Furthermore, this would looks redundant when you try to call it: `imotioncube.resetIMotionCube()`.

File: march_hardware/src/Joint.cpp [hunk 21]
```diff
@@ -55,6 +55,14 @@ void Joint::prepareActuation()
   ROS_INFO("[%s] Successfully prepared for actuation", this->name_.c_str());
 }
 
+void Joint::resetIMotionCube()
+{
+  if (hasIMotionCube())
```
Comment (Olavhaasie) [inline march_hardware/src/Joint.cpp]:
```suggestion
  if (this->hasIMotionCube())
```

File: march_hardware_interface/include/march_hardware_interface/march_hardware_interface.h [hunk 22]
```diff
@@ -34,7 +34,7 @@ using RtPublisherPtr = std::unique_ptr<realtime_tools::RealtimePublisher<T>>;
 class MarchHardwareInterface : public hardware_interface::RobotHW
 {
 public:
-  explicit MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot);
+  explicit MarchHardwareInterface(std::unique_ptr<march::MarchRobot> robot, bool reset_imc);
```
Comment (Olavhaasie) [inline march_hardware_interface/include/march_hardware_interface/march_hardware_interface.h]: `explicit` is no longer necessary, since that is only valid for single argument constructors.

**Review by Olavhaasie — CHANGES_REQUESTED** [review]

File: march_hardware/include/march_hardware/IMotionCube.h (hunk 20 above)
Comment (Joris-JW) [reply march_hardware/include/march_hardware/IMotionCube.h]: right, good call. Nice review!

File: march_hardware_interface/launch/hardware.launch (hunk 17 above)
Comment (Roelemans) [reply march_hardware_interface/launch/hardware.launch]: Sorry this comment was not very clear. I meant remove empty line 3 and empty line 5. That way, the args are grouped, and seperated from the rest by a single newline. 

tl;dr remove empty line 3 too.

File: march_hardware_interface/launch/hardware.launch [hunk 23]
```diff
@@ -1,6 +1,8 @@
 <launch>
     <arg name="robot" default="march4" doc="The robot to run. Can be: march3, march4, test_joint_linear, test_joint_rotational."/>
 
+    <arg name="reset_imc" default="false" doc="Reset the IMC if this argument is set to true"/>
+
```
Comment (Roelemans) [inline march_hardware_interface/launch/hardware.launch]:
```suggestion
<launch>
    <arg name="robot" default="march4" doc="The robot to run. Can be: march3, march4, test_joint_linear, test_joint_rotational."/>
    <arg name="reset_imc" default="false" doc="Reset the IMC if this argument is set to true"/>

```
Sorry the comment here was not very clear. I meant remove empty line 3 and empty line 5. That way, the args are grouped, and separated from the rest by a single newline.

**Review by Roelemans — CHANGES_REQUESTED** [review]

**Review by Roelemans — DISMISSED** [review]
Nice testing :cat:

**Review by Olavhaasie — DISMISSED** [review]

**Review by Roelemans — APPROVED** [review]

**Review by Olavhaasie — APPROVED** [review]
